src/CommonLib/Ntlm/LdapNative.cs

```diff
@@ -0,0 +1,346 @@
+            Marshal.Copy(responseBerval.bv_val, responseData, 0, responseBerval.bv_len);
+            return responseData;
+        }
+        finally
+        {
+            if (credPtr != IntPtr.Zero)
+            {
+                var berval = Marshal.PtrToStructure<BERVAL>(credPtr);
+                if (berval.bv_val != IntPtr.Zero)
+                {
+                    Marshal.FreeHGlobal(berval.bv_val);
+                }
+                Marshal.FreeHGlobal(credPtr);
+            }
+
+            // Freeing the memory always results in crashes. Is it managed by something else?
+            //if (response != IntPtr.Zero)
+            //{
+            //    var responseBerval = Marshal.PtrToStructure<NativeMethods.berval>(response);
+            //    if (responseBerval.bv_val != IntPtr.Zero)
+            //    {
+            //        Marshal.FreeHGlobal(responseBerval.bv_val);
+            //    }
+            //    Marshal.FreeHGlobal(response);
+            //}
+        }
+    }
+    public void Dispose()
+    {
+        if (!_disposed)
+        {
+            _handle?.Dispose();
+            _handle = null;
+            _disposed = true;
+        }
+    }
+}
+
+
+[StructLayout(LayoutKind.Sequential)]
+public struct LDAP_TIMEVAL
+{
+    public int tv_sec;
+    public int tv_usec;
+}
+
+[StructLayout(LayoutKind.Sequential)]
+public struct BERVAL
+{
+    public int bv_len;
+    public IntPtr bv_val;
+}
+public class NativeMethods
+{
+
+    [DllImport("Wldap32.dll", CallingConvention = CallingConvention.Cdecl)]
+    public static extern IntPtr ldap_init(string Host, int Port);
+
+    [DllImport("Wldap32.dll", CallingConvention = CallingConvention.Cdecl)]
+    public static extern int ldap_connect(SafeLdapHandle ld, IntPtr timeout);
+
+    [DllImport("Wldap32.dll", CallingConvention = CallingConvention.Cdecl)]
+    public static extern int ldap_set_option(SafeLdapHandle ld, LdapOption option, ref int value);
+
+    [DllImport("Wldap32.dll", CallingConvention = CallingConvention.Cdecl)]
+    public static extern int ldap_set_option(SafeLdapHandle ld, LdapOption option, IntPtr value);
+
+    [DllImport("Wldap32.dll", CallingConvention = CallingConvention.Cdecl)]
+    public static extern int ldap_get_option(SafeLdapHandle ld, LdapOption option, out int value);
+    [DllImport("Wldap32.dll", CallingConvention = CallingConvention.Cdecl)]
+    public static extern int ldap_get_option(SafeLdapHandle ld, LdapOption option, out IntPtr value);
+
+    [DllImport("Wldap32.dll", CallingConvention = CallingConvention.Cdecl)]
+    public static extern void ldap_memfree(IntPtr ptr);
+
+    [DllImport("Wldap32.dll", EntryPoint = "ldap_sasl_bind_s", CallingConvention = CallingConvention.Cdecl)]
+    public static extern int ldap_sasl_bind_s(
+        SafeLdapHandle ld,
+        string dn,
+        string mechanism,
+        IntPtr cred,
+        IntPtr serverctrls,
+        IntPtr clientctrls,
+        out IntPtr msgidp);
```

Comment:
Nit: this name doesn't match the output defined in the library
https://learn.microsoft.com/en-us/windows/win32/api/winldap/nf-winldap-ldap_sasl_bind_sa

Reply:
Unfortunately, pinvoke is weird, this is normal